Add support for PER_NODE capacity on Scheduled Executor

tkountis opened the pull request:
Introduce capacity policy per executor, which allows users to declare capacity PER_NODE or PER_PARTITION (old way). The new default is PER_NODE, for ease of use.

**Important note:** migrations are excluded of this policy, to prevent any undesirable data-loss. The migrated tasks will be accounted for, but they wont be rejected. 

Fixes https://github.com/hazelcast/hazelcast/issues/11629

vojtechtoman commented on hazelcast/src/main/resources/hazelcast-full-example.xml:
```diff
@@ -881,6 +881,7 @@
         <pool-size>16</pool-size>
         <durability>1</durability>
         <capacity>100</capacity>
+        <capacity-policy>PER_NODE</capacity-policy>
```
Missing documentation in the `scheduled-executor-service` comment block above.

vojtechtoman commented on hazelcast/src/test/java/com/hazelcast/scheduledexecutor/ScheduledExecutorServiceBasicTest.java:
```diff
@@ -155,7 +177,34 @@ public void capacity_whenDefault() {
+            fail("Should have been rejected.");
+        } catch (RejectedExecutionException ex) {
+            assertEquals("Got wrong RejectedExecutionException",
+                    "Maximum capacity (100) of tasks reached for partition (262) and scheduled executor (foobar). "
```
What about replacing the constant 262 with something like `getNodeEngine().getPartitionService().getPartitionId(keyOwner)`?

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/config/ScheduledExecutorConfig.java:
```diff
@@ -56,6 +57,8 @@
 
     private int capacity = DEFAULT_CAPACITY;
 
+    private CapacityPolicy capacityPolicy = PER_NODE;
```
Missing in `writeObject`/`readObject`, copy constructor and `toString`

mmedenjak replied:
Is it just me or is it still missing? :)

mmedenjak replied:
Khm khm :)

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/AbstractScheduledExecutorContainerHolder.java:
```diff
@@ -32,14 +32,16 @@
         implements ScheduledExecutorContainerHolder {
 
     final NodeEngine nodeEngine;
+    final TaskLifecycleHook taskLifecycleHook;
```
This field can be avoided if the `ScheduledExecutorMemberBin` initialises the constructor function in the constructor, like `ScheduledExecutorPartition` does (and you get a nice lambda + remove `nodeEngine` field)

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/ScheduledExecutorMemberBin.java:
```diff
@@ -16,14 +16,19 @@
+     * Capacity controlled by PER_NODE config in the service layer
+     * (if set, {@link DistributedScheduledExecutorService#ensurePerNodeCapacity(String)}).
+     */
+    private static final int CAPACITY_DISABLED = 0;
```
Maybe make this a constant on some shared class so `ScheduledExecutorPartition#ScheduledExecutorPartition` and `ScheduledExecutorContainer#checkNotAtCapacity` can use it as a constant.

mmedenjak commented:
Another design question/alternative - what if we provided the `AtomicLong` as the "permit count" directly to the containers and have them enforce the capacity?  In the case of per-partition, the `AtomicLong` could be initialised to the per-partition capacity and a new instance would be created per container while for a per-node capacity, all containers would share a single `AtomicLong`.

tkountis commented:
Brilliant idea @mmedenjak (why o' why didn't i think of that :) )
I changed the implementation, it was a much cleaner approach than the lifecycle hook.

tkountis commented on hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/ScheduledExecutorMemberBin.java:
```diff
@@ -21,30 +21,39 @@
+
+    CapacityPermit newPermitFor(String name, DistributedScheduledExecutorService service,
+                                ScheduledExecutorConfig config) {
+        if (config.getCapacity() == 0 || config.getCapacityPolicy() == PER_PARTITION) {
```
Previous behavior was treated as partition, but I think its misleading to use the scheduleOnMember API and have PER_PARTITION restrictions at this stage. I documented it on the config.

mmedenjak replied:
👍

mmedenjak replied:
👍

mmedenjak commented:
One more question - we release the permit once a task is disposed, right? So if the task has already run and completed, we still hold a permit?
What if a completed task is migrated - does it acquire a permit on the new partition owner? Or do we acquire a task only if the task has not yet completed and we schedule it to run on the new owner?

tkountis commented:
Permits are equal to the number of tasks stored in the partition(s). It only gets releases once the task is removed (ie. disposed). Completed tasks still acquire resources until the user disposes them, so the restriction is based on the allocated resources and not on the meaning of completion. Migrated tasks work the same, they acquire the permit even if they are completed.

tkountis commented:
> Looks good! Added some minor points for consideration. Another thing (driven by experience with WAN queue counters) - do we handle migration disposing the capacity when the partition owner changes? I think we should handle it in `ScheduledExecutorPartition#disposeObsoleteReplicas`. And can you add a test case for it, like adding and removing members? I just have such bad experience with WAN counters being wrong :)

Capacity is released upon `ScheduledExecutorPartition#disposeObsoleteReplicas` which calls `ScheduledExecutorContainer#destroy` which disposes all tasks in a loop, releasing the permits.
You are right about a test, thought i forgot to add one.

tkountis commented:
@mmedenjak have another look, addressed your comments, and bases on your suggested test I found an issue with replicas counting. The logic was changed a bit, now the counting is enforced through scheduling and not-enforced through promotion. Release is happening through disposal and suspend.

tkountis commented:
Unrelated test-failures

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/ScheduledTaskDescriptor.java:
```diff
@@ -122,16 +122,21 @@ Object get()
+     * it will be assigned on a different Future. <p> Task ownership is also restored to default,
      * which will be fixed when migrations finish and a new master is selected.
+     *
+     * @return <code>true</code> if the task was actually scheduled
```
... suspended

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/ScheduledExecutorContainer.java:
```diff
@@ -228,9 +251,16 @@ public void promoteSuspended() {
+                    acquirePermit(true);
+                }
+
+                descriptor.setActive();
```
Doesn't `doSchedule(descriptor);` do the same thing in `descriptor.setScheduledFuture(future);`?

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/ScheduledExecutorContainer.java:
```diff
@@ -228,9 +251,16 @@ public void promoteSuspended() {
                 }
+
+                if (!wasActive) {
+                    acquirePermit(true);
```
Can we move this inside `if (descriptor.canBeScheduled())`?

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/ScheduledTaskDescriptor.java:
```diff
@@ -98,22 +103,24 @@ void setState(Map<?, ?> snapshot) {
-    void setTaskResult(ScheduledTaskResult result) {
-        this.resultRef.set(result);
+    boolean isActive() {
+        return this.status.get() == ACTIVE;
```
Generally speaking, what if we didn't have this additional status and we relied on the results of the old `ScheduledTaskDescriptor#canBeScheduled` implementation?

tkountis replied:
Will merge this PR. There is an on-going discussion on slack regarding this, based on the outcome we can open another PR to address it if needed.
